### ISSUE TITLE
Nice to have: don't connect to X from init().

### ISSUE THREAD

**Issue by divVerent** [hajimehoshi/ebiten#1498]
Currently, ebiten connects to X right at startup, even before main() runs:

```
$ DISPLAY= ./aaaaaa-debug-linux-amd64                                                               
2021/02/17 10:33:16 PlatformError: X11: Failed to open display 
panic: NotInitialized: The GLFW library is not initialized

goroutine 1 [running]:
github.com/go-gl/glfw/v3.3/glfw.acceptError(0x0, 0x0, 0x0, 0x0, 0x10006)
	/home/rpolzer/go/pkg/mod/github.com/go-gl/glfw/v3.3/glfw@v0.0.0-20201108214237-06ea97f0c265/error.go:174 +0x225
github.com/go-gl/glfw/v3.3/glfw.panicError(...)
	/home/rpolzer/go/pkg/mod/github.com/go-gl/glfw/v3.3/glfw@v0.0.0-20201108214237-06ea97f0c265/error.go:185
github.com/go-gl/glfw/v3.3/glfw.WindowHint(0x20004, 0x0)
	/home/rpolzer/go/pkg/mod/github.com/go-gl/glfw/v3.3/glfw@v0.0.0-20201108214237-06ea97f0c265/window.go:275 +0x49
github.com/hajimehoshi/ebiten/v2/internal/glfw.WindowHint(...)
	/home/rpolzer/go/pkg/mod/github.com/hajimehoshi/ebiten/v2@v2.0.5/internal/glfw/glfw_notwindows.go:329
github.com/hajimehoshi/ebiten/v2/internal/uidriver/glfw.initialize(0xc00009a690, 0xc00009a660)
	/home/rpolzer/go/pkg/mod/github.com/hajimehoshi/ebiten/v2@v2.0.5/internal/uidriver/glfw/ui.go:133 +0x4f
github.com/hajimehoshi/ebiten/v2/internal/uidriver/glfw.init.1()
	/home/rpolzer/go/pkg/mod/github.com/hajimehoshi/ebiten/v2@v2.0.5/internal/uidriver/glfw/ui.go:120 +0x25
```

This makes it impossible to write headless unit tests for parts of a module that should in theory be testable; one would then have to fully separate testable code from ebiten-using code into separate packages, rather than just e.g. mocking out uses of ebiten.Image and similar stuff at runtime.

It'd be nice to instead to connect to the windowing system only once needed, i.e. once some functions are called that really need it (e.g. RunGame, DeviceScaleFactor).

Yes, would probably be rather duplicative to add this now without introducing a new ebiten.Init() function call as you probably want to stay compatible within 2.x. Although, maybe we could at least have a build tag to turn off this init() handling and rather require the main function to call it explicitly before using ebiten?

**Comment by hajimehoshi** [comment]
There is an issue for the headless mode (#353). This is not an easy task to implement but should be useful.

**Comment by divVerent** [comment]
I understand, but I do not actually need that much - just that code that
does not call ebiten but just uses its types can be compiled and run
without X.

(E.g. in my case I could then replace my image loading by only ever
returning nil as *ebiten.Image, implement but never call Draw, have custom
wrappers for the other ebiten global functions e.g. for input, and I could
run my game headless easily)

If I were to send a PR for this (basically adding an opt-in build tag to
make the init call explicit, with the goal to make this call required in
ebiten v3), would you accept it? Again, this would not be a real headless
mode, and I would expect many ebiten calls to fail more or less violently
before Init, but it would e.g. allow for what Quake does, having headless
game server (that does not use ebiten) and graphical game client in the
same binary, decided by command line flag.

**Comment by hajimehoshi** [comment]
> (E.g. in my case I could then replace my image loading by only ever
returning nil as *ebiten.Image, implement but never call Draw, have custom
wrappers for the other ebiten global functions e.g. for input, and I could
run my game headless easily)

I'd make a wrapper for *ebiten.Image and other Ebiten APIs so that your library doesn't depend on Ebiten directly.

IMO it is not feasible to ensure that Ebiten's initialization logic doesn't cause any side effects like X now and in the future, and I tend not to want to introduce a build tag for this purpose.
